java/ql/src/semmle/code/java/frameworks/JaxWS.qll

```diff
@@ -955,3 +961,17 @@ private class VulnerableEntity extends XssSinkBarrier {
       ).getArgument(0)
   }
 }
+
+/**
+ * Model sources stemming from `ContainerRequestContext`.
+ */
+private class ContainerRequestContextModel extends SourceModelCsv {
+  override predicate row(string s) {
+    s =
+      ["javax", "jakarta"] + ".ws.rs.container;ContainerRequestContext;true;" +
+        [
+          "getAcceptableLanguages", "getAcceptableMediaTypes", "getCookies", "getEntityStream",
+          "getHeaders", "getHeaderString", "getUriInfo"
+        ] + ";;;ReturnValue;remote"
```

Comment:
Not sure how feasible that is right now but I wonder if the context should also be treated like a Map to ensure the `setProperty` and `getProperty` calls propagate taint.

Reply:
In common with my approach to JSON-related libs I didn't bother to distinguish them because the use case is deserialization -> read or write -> serialization, where we only care about a single monolithic concept of taint, as opposed to map-writes followed by map-reads as with a general-purpose data structure. If we catch people actually using these like general-purpose data structures in the wild though, then we should do the extra work and increase the fidelity of our modelling as you suggest.

Reply:
Generally agree. In this particular case, I thought that the context properties are actually the designated way to pass data among the filters and to the content producers (see also https://abhirockzz.wordpress.com/2016/03/16/sharing-data-between-jax-rs-filters/ ). But this may be better modelled as part of a query than here.

Reply:
Ah, interesting, hadn't seen that done. To support that case we would need to model the control-flow from one filter to another, since that is implemented in opaque framework code.